### DISCUSSION TITLE
Prometheus explorer / autocompleter user experience and performance

### DISCUSSION THREAD

**Discussion by aughr** [grafana/grafana#46432] [Observability]
Hey Grafana community, I work at AWS on our managed Grafana and Prometheus services. We've noticed that our larger customers have run into performance and reliability issues when interacting with their Prometheus (Cortex) data in Grafana over very large time windows. Customers see the autocompleter-driven queries timing out or triggering slowness in their browser, resulting in being unable to explore their data properly.

As far as I can tell, we have a few different issues:
- `/api/v1/labels`, `/api/v1/label/<label_name>/values`, and `/api/v1/series` are expected to return a potentially unbounded set of results. They take input parameters to limit the data that's queried, but for example there's nothing stopping a customer with a large number of series for a given metric name returning hundreds of MBs of data. In larger scale deployments like on Cortex, too, we may want to limit the data we scan in any given metadata request, but currently we have no way to signal to the client that we either limited the returned data and/or paginated it.
- As mentioned in https://github.com/grafana/grafana/issues/33487, the existing APIs could be used more efficiently. I don't know how far we could go here in optimizing the calls to Prometheus.

As @bboreham mentioned on that linked issue, one blocker here is that Prometheus can't tell Grafana what features it supports, so any improvements we make will be blocked on [a features endpoint or some other solution](https://github.com/prometheus/prometheus/issues/10022).

I'd like to work backwards from the user experience we want in Grafana before taking this to a design doc / pull request in Prometheus, but I do expect we'll end up wanting a Prometheus API change and then a corresponding Cortex change in the end.

Here's my current thinking: we should prioritize interactivity in the Grafana UI's autocompletion and explorer views with the ability to scan progressively more data for a needle in a haystack if desired. The initial clicks to find metric names or complete label names or values for a series should be fast. If we can get all the data we need, great, but otherwise we should know that there's potentially more data to be had. We may or may not need to want to actually fetch that data versus waiting for the user to refine their search via more characters or labels in a filter.

A few implementation options, not mutually exclusive, that come to mind:
- Even when the explorer / query editor time window is extremely long, only query for a small window of the more recent data. Indicate to the user that we can try for more, then page through time.
- We could add a flag to the Prometheus API to say we want to prioritize fast responses, not scanning all the data if it's too large, e.g. in a large Cortex installation. Prometheus could then respond with whether the data is complete or not. This might be coupled with cursor-based pagination, and/or…
- In the face of an incomplete response from Prometheus, we could query Prometheus with a more specific query, e.g. a metric name prefix.

Thoughts on the problem, the UX, and/or the implementations?

**Comment by aocenas** [comment]
Probably one option would also be to push the query and processing of this into a webworker so at least it does not block even if it isn't well optimized.

**Reply by aughr** [reply]
Oh yeah, that makes a lot of sense. Is there a pattern in the Grafana codebase for that already?

**Reply by aocenas** [reply]
Only thing that I know of is NodeGraph visualisation that does computing of the node layout in web worker https://github.com/grafana/grafana/blob/d87cd6f26c82000f7dadaa47577eb139853df1da/public/app/plugins/panel/nodeGraph/layout.ts#L165

**Reply by gabor** [reply]
the web-worker can help, but i recommend measuring it first... when the monaco-autocomplete happens, it's a promise-returning-callback, so it does not block while we wait for the ajax-response.  the ajax_response_to_list_of_autocomplete_items processing does block, so webworker-ing could help there.

**Comment by davkal** [comment]
I quite like the windowed approach for queries. The problem has been that we can't know how big the prom/cortex installation really is, until we query it. Even for long time ranges the amount of data can vary wildly, but slicing seems a reasonable heuristic.

**Comment by taleena** [comment]
Ha! You found it before I even pinged you, wonderful. :) @RichiH you were also mentioned as potentially being interested in this one.

**Comment by bboreham** [comment]
Note that autocomplete is not the only source of such requests.
Many people make dashboards with drop-downs to select context at the top, like this:
![image](https://user-images.githubusercontent.com/8125524/158356266-3f3d24a4-70ad-4031-858a-91dd7cfe0754.png)
(this is from [here](https://grafana.com/blog/2021/12/02/new-in-the-kubernetes-integration-for-grafana-cloud-curated-dashboards-built-in-alerts-and-more/))

This is different to autocomplete: while the UI still works if an autocomplete request is abandoned, these variables at the top must be populated. If a user extends the time range of the dashboard out to 30 or 90 days, it is a lot of work in the back end to come back with the complete range of options, particularly when Grafana is using `/series` to fetch label-values.

I wanted to mention this because I think it's a similar problem that needs a different approach.  Maybe Grafana can add options to say "only query the last 24 hours of the range to populate variables", or grey out the variable choices when the range is very long.

**Reply by aughr** [reply]
Thanks @bboreham, that's a great call-out. In addition to the Grafana-side option, what do you think of the idea of having Prometheus/Cortex limit and/or paginate responses when the data set to scan or return is just too big?

**Reply by bboreham** [reply]
Pagination doesn't seem right.  For both autocomplete and drop-down variables the client never wants a lot of data - a hundred replies at most.

Limiting the replies would help the client, but I would also want to help the server - scanning 90 days of index covering 10 million series is always going to be a lot of work, even if it comes back with only 2 values.  So I want to limit the time range queried.

**Comment by bboreham** [comment]
Small update: 

> Prometheus can't tell Grafana what features it supports

Grafana now [has this ability](https://grafana.com/docs/grafana/latest/datasources/prometheus/#configure-the-data-source): "When selected, the Prometheus version field attempts to detect the version automatically using the Prometheus buildinfo API."

**Comment by itsmylife** [comment]
Most of the requests have been handled already and we will keep making improvements. 

Currently,
- We have been running UX research to see the impactful changes
- Improving variable editor
- Working on decoupling Prometheus from the core Grafana so provide more granular updates

We welcome all the ideas within our development team to continue making improvements, so please keep sharing!
